windows/src/support/oskbulkrenderer/Keyman.System.OSKBulkRenderer.pas

```diff
@@ -0,0 +1,267 @@
+    then destfilename := IncludeTrailingPathDelimiter(FTargetFolder) + ChangeFileExt(ExtractFileName(srcfilename), '.png')
+    else destfilename := ChangeFileExt(srcfilename, '.png');
+
+  writeln(srcfilename + ' => ' + destfilename);
+
+  vk := TVisualKeyboard.Create;
+  try
+    vk.LoadFromFile(srcfilename);
+
+    if FTargetFolder <> '' then
+    begin
+      WriteIndex('<h2>'+srcfilename+'</h2>');
+      WriteIndex('<dl>');
+      WriteIndex('  <dt>Font</dt><dd>'+vk.Header.UnicodeFont.Name+'</dd>');
+      s := '';
+      if kvkh102 in vk.Header.Flags then s := s + 'key102 ';
+      if kvkhDisplayUnderlying in vk.Header.Flags then s := s + 'displayunderlying ';
+      if kvkhUseUnderlying in vk.Header.Flags then s := s + 'useunderlying ';
+      if kvkhAltGr in vk.Header.Flags then s := s + 'usealtgr ';
+      WriteIndex('  <dt>Flags</dt><dd>'+Trim(s)+'</dd>');
+
+      if vk.Header.UnderlyingLayout <> '' then
+        WriteIndex('  <dt>Base Layout</dt><dd>'+vk.Header.UnderlyingLayout+'</dd>');
+
+      s := '';
+      for FShift := 0 to High(VKLegalShiftStates) do
+      begin
+        if vk.HasShiftState(VKLegalShiftStates[FShift].Shift) then
+        begin
+          if s <> '' then s := s + ', ';
+          s := s + VKLegalShiftStates[FShift].Desc;
+        end;
+      end;
+      WriteIndex('  <dt>Shift States</dt><dd>'+s+'</dd>');
+      WriteIndex('  <dt>Associated Keyboard</dt><dd>'+vk.Header.AssociatedKeyboard+'</dd>');
+      WriteIndex('</dl>');
+      WriteIndex('<div><img src="'+ExtractFileName(destfilename)+'" alt="'+ExtractFileName(destfilename)+'"></div>');
+      WriteIndex('');
+    end;
+
+    if FileExists(destfilename) and not FOverwrite then
+    begin
+      writeln('File '+destfilename+' already exists. Skipping');
+      Exit(True);
+    end;
+
+
+    vkep := TVisualKeyboardExportPNG.Create(vk, False, False, True, FWidth);
+    try
+      vkep.ExportToFile(destfilename);
+    finally
+      vkep.Free;
+    end;
+  finally
+    vk.Free;
+  end;
+
+  Result := True;
+end;
+
+class procedure TOSKBulkRenderer.WriteHTMLHeader;
+begin
+  AssignFile(FIndexFile, IncludeTrailingPathDelimiter(FTargetFolder) + 'index.html');
+  Rewrite(FIndexFile);
+  WriteIndex('<!DOCTYPE HTML>');
+  WriteIndex('<html><head>');
+  WriteIndex('  <meta charset="utf-8">');
+  WriteIndex('  <title>On Screen Keyboard Renders</title>');
+  WriteIndex('</head>');
+  WriteIndex('<body>');
+  WriteIndex('<h1>On Screen Keyboard Renders</h1>');
+  WriteIndex('<dl>');
+  WriteIndex('  <dt>Date</dt><dd>'+FormatDateTime('yyyy-mm-dd hh:nn:ss', Now)+'</dd>');
+  WriteIndex('  <dt>Version</dt><dd>'+GetVersionString+'</dd>');
+  WriteIndex('</dl>');
+  WriteIndex('');
+end;
+
+class procedure TOSKBulkRenderer.WriteHTMLFooter;
+begin
+  WriteIndex('</body></html>');
+  CloseFile(FIndexFile);
+end;
+
+class procedure TOSKBulkRenderer.WriteIndex(s: string);
+var
+  us: utf8string;
+begin
+  us := UTF8String(s);
+  writeln(FIndexFile, us);
+end;
+
+end.
```

Comment:
lgtm. This diff goes away w/ a rebase, right?

Reply:
Indeed. See now :)